edx_proctoring/api.py

```diff
@@ -594,113 +692,26 @@ def create_exam_attempt(exam_id, user_id, taking_as_proctored=False):
-        # exceptions are granted on a individual basis as an
-        # allowance
-        if review_policy_exception:
-            context.update({
-                'review_policy_exception': review_policy_exception
-            })
-
-        # now call into the backend provider to register exam attempt
-        try:
-            external_id = get_backend_provider(exam).register_exam_attempt(
-                exam,
-                context=context,
-            )
-        except BackendProviderSentNoAttemptID as ex:
-            log_message = (
-                u'Failed to get the attempt ID for {user_id}'
-                u'in {exam_id} from the backend because the backend'
-                u'did not provide the id in API response, even when the'
-                u'HTTP response status is {status}, '
-                u'Response: {response}'.format(
-                    user_id=user_id,
-                    exam_id=exam_id,
-                    response=str(ex),
-                    status=ex.http_status
-                )
-            )
-            log.error(log_message)
-            raise ex
-        except BackendProviderCannotRegisterAttempt as ex:
-            log_message = (
-                u'Failed to create attempt for {user_id} '
-                u'in {exam_id} because backend was unable '
-                u'to register the attempt. Status: {status}, '
-                u'Reponse: {response}'.format(
-                    user_id=user_id,
-                    exam_id=exam_id,
-                    response=str(ex),
-                    status=ex.http_status,
-                )
-            )
-            log.error(log_message)
-            raise ex
-        except BackendProviderOnboardingException as ex:
-            force_status = ex.status
-            log_msg = (
-                u'Failed to create attempt for {user_id} '
-                u'in {exam_id} because of onboarding failure: '
-                u'{force_status}'.format(**locals())
-            )
-            log.error(log_msg)
-
-    if not external_id and exam['backend'] == 'proctortrack' and taking_as_proctored:
-        # temporary logging added to better understand cases where we end up
-        # with exam attempts with NULL external_ids - a situation, that, theoretically,
-        # shouldn't be occurring (EDUCATOR-4948)
+    if is_exam_passed_due(exam, user=user_id) and taking_as_proctored:
         log_msg = (
-            u'Registered exam attempt with an external ID {external_id}'
-            u'in {exam_id} for user {user_id}'.format(
-                external_id=external_id,
-                exam_id=exam_id,
-                user_id=user_id,
+            u'user_id {user_id} trying to create exam atttempt for past due proctored exam {exam_id} '
+            u'Do not register an exam attempt! Return None'.format(
+                exam_id=exam_id, user_id=user_id
             )
         )
         log.error(log_msg)
+        return None
```

Comment:
I would prefer that we threw an exception here, since we know we're in an error situation at this point. I think this is using a return value for some error cases, and exceptions for others. I would prefer if we were consistent (especially in the absence of a note in the docstring).

Reply:
@matthugs Got it. I updated the code according to your feedback and now we are raising exception.